src/ucp/proto/proto_perf.c

```diff
@@ -54,50 +54,52 @@ struct ucp_proto_perf {
-    [UCP_PROTO_PERF_FACTOR_LOCAL_CPU]            = "cpu",
-    [UCP_PROTO_PERF_FACTOR_REMOTE_CPU]           = "cpu-remote",
-    [UCP_PROTO_PERF_FACTOR_LOCAL_TL]             = "tl",
-    [UCP_PROTO_PERF_FACTOR_REMOTE_TL]            = "tl-remote",
-    [UCP_PROTO_PERF_FACTOR_LOCAL_MEMTYPE_COPY]   = "memtype-copy",
-    [UCP_PROTO_PERF_FACTOR_REMOTE_MEMTYPE_COPY]  = "memtype-copy-remote",
-    [UCP_PROTO_PERF_FACTOR_LATENCY]              = "lat",
-    [UCP_PROTO_PERF_FACTOR_LAST]                 = NULL
+    [UCP_PROTO_PERF_FACTOR_LOCAL_CPU]     = "cpu",
+    [UCP_PROTO_PERF_FACTOR_REMOTE_CPU]    = "cpu-remote",
+    [UCP_PROTO_PERF_FACTOR_LOCAL_TL]      = "tl",
+    [UCP_PROTO_PERF_FACTOR_REMOTE_TL]     = "tl-remote",
+    [UCP_PROTO_PERF_FACTOR_LOCAL_MTCOPY]  = "mtcopy",
+    [UCP_PROTO_PERF_FACTOR_REMOTE_MTCOPY] = "mtcopy-remote",
+    [UCP_PROTO_PERF_FACTOR_LATENCY]       = "lat",
+    [UCP_PROTO_PERF_FACTOR_LAST]          = NULL
 };
 
-#if ENABLE_ASSERT
-static const char*
-ucp_proto_perf_log(const ucp_proto_perf_t *perf, ucs_log_level_t log_level)
-{
-    ucs_string_buffer_t strb = UCS_STRING_BUFFER_INITIALIZER;
-
-    ucp_proto_perf_str(perf, &strb);
-    ucs_log(log_level, "%s %s", perf->name, ucs_string_buffer_cstr(&strb));
-    ucs_string_buffer_cleanup(&strb);
-    return "";
-}
-#endif
-
 static void ucp_proto_perf_check(const ucp_proto_perf_t *perf)
 {
 #if ENABLE_ASSERT
+    ucs_string_buffer_t funcb = UCS_STRING_BUFFER_INITIALIZER;
+    const char *reason        = NULL;
     const ucp_proto_perf_segment_t *seg;
     size_t min_start;
 
     min_start = 0;
     ucp_proto_perf_segment_foreach(seg, perf) {
-        ucs_assertv((seg->start >= min_start) && (seg->start <= seg->end),
-                    "perf=%p seg->start=%zu seg->end=%zu min_start=%zu %s",
-                    perf, seg->start, seg->end, min_start,
-                    ucp_proto_perf_log(perf, UCS_LOG_LEVEL_ERROR));
-        if (seg->end == SIZE_MAX) {
-            ucs_assertv(ucs_list_is_last(&perf->segments, &seg->list),
-                        "perf=%p seg->start=%zu seg->end=%zu %s",
-                        perf, seg->start, seg->end,
-                        ucp_proto_perf_log(perf, UCS_LOG_LEVEL_ERROR));
-        } else {
+        if (seg->start < min_start) {
+            reason = "seg->start < min_start";
+            break;
+        }
+        if (seg->start > seg->end) {
+            reason = "seg->start > seg->end";
+            break;
+        }
+        if (seg->end < SIZE_MAX) {
             min_start = seg->end + 1;
+        } else {
+            if (!ucs_list_is_last(&perf->segments, &seg->list)) {
+                reason = "!ucs_list_is_last(&perf->segments, &seg->list)";
+                break;
+            }
         }
     }
+
+    if (reason != NULL) {
+        ucp_proto_perf_str(perf, &funcb);
+        ucs_error("%s %s", perf->name, ucs_string_buffer_cstr(&funcb));
+        ucs_fatal("%s perf=%p seg->start=%zu seg->end=%zu min_start=%zu",
```

Comment:
why need both ucs_error and ucs_fatal?

Reply:
I made it because printing whole perf structure information inside `ucs_fatal` message can be overwhelming FMPOV. I changed `ucs_error` to `ucs_debug` to hide this information from user when they get issue first time, but still save ability to print the whole structure from release binary.